kube-runtime/src/reflector/shared_store.rs

```diff
@@ -0,0 +1,401 @@
+use crate::{
+    reflector::{
+        reflector,
+        store::{Store, Writer},
+    },
+    utils::{stream_subscribe, StreamSubscribe, WatchStreamExt},
+    watcher::{self, watcher, Event},
+};
+use futures::{stream, Stream, StreamExt};
+use k8s_openapi::NamespaceResourceScope;
+use kube_client::{api::ListParams, Api, Client, Resource};
+use serde::de::DeserializeOwned;
+use std::{collections::HashMap, fmt::Debug, hash::Hash, sync::Arc};
+use stream::BoxStream;
+
+pub struct SharedStore<K, W = WatcherFactory<K>>
+where
+    K: 'static + Resource + Clone + DeserializeOwned + Debug + Send,
+    K::DynamicType: Hash + Eq,
+    W: CreateWatcher<K>,
+{
+    watcher_provider: W,
+    reflectors: HashMap<ScopedListParams, (Store<K>, BoxStreamSubscribe<K>)>,
+}
+
+type ScopedListParams = (Option<String>, ListParams);
+type BoxStreamSubscribe<K> = StreamSubscribe<BoxStream<'static, watcher::Result<Event<K>>>>;
+
+impl<K> SharedStore<K>
+where
+    K: 'static + Resource + Clone + DeserializeOwned + Debug + Send + Sync,
+    K::DynamicType: Default + Eq + Hash + Clone,
+{
+    pub fn new(client: Client) -> Self {
+        Self {
+            watcher_provider: WatcherFactory::new(client),
+            reflectors: HashMap::new(),
+        }
+    }
+}
+
+impl<K, W> SharedStore<K, W>
+where
+    K: Clone + Resource + DeserializeOwned + Debug + Send + Sync + 'static,
+    K::DynamicType: Default + Eq + Hash + Clone,
+    W: CreateWatcher<K> + 'static,
+{
+    pub fn run(self) -> impl Stream<Item = Arc<watcher::Result<Event<K>>>> {
+        stream::select_all(self.reflectors.into_iter().map(|(_, (_, reflector))| reflector))
+    }
+
+    pub fn namespaced(&mut self, namespace: &str, list_params: ListParams) -> Store<K>
+    where
+        K: Resource<Scope = NamespaceResourceScope>,
+    {
+        self.reflectors
+            .entry((Some(namespace.to_string()), list_params.clone()))
+            .or_insert_with(|| setup_reflector(self.watcher_provider.namespaced(namespace, list_params)))
+            .0
+            .clone()
+    }
```

Comment:
The `Scope` constraints on the `Resource` are nice ways of preventing misuse, but they are just deferring to similarly scoped `Api` helpers below, and effectively duplicating a lot of its setup to achieve the same constraints. 

It might not be anything wrong with that, but when analysing this I was wondering if it made sense to drop the `WatcherFactory` entirely and create a simpler `From Api<K> for SharedStore<K>` instead so that we don't have to think about scope and defer to the scope of the `Api` instead. You'd end up with a much cleaner interface, but you also have a `SharedStore` that is bound to a `Scope` and you need multiple ones if you need one controller looking at entries at the global level and another at a namespace level.

...It's possible that hardcoding a scope like this this can be a good thing. If you say `::namespaced` then you will only ever get subsets of things from one namespace, but if you say `::all` you gain basically all the flexibility; through `ListParams` you could restrict what namespaces (using [chained field-selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/#chained-selectors)) you actually want to consider (and if it's cluster scoped then you'll always want to use cluster scope anyway).

What do you think about this line? This type of setup could provide a better starting point for us to better allow intelligent watcher reduction without duplicating the IO. I'll write up an issue on how I think this can go - and if we are careful - we can maybe avoid solving it all here.

Reply:
ok. this is not as easy as i thought when writing this comment, but there is a possible way forward. I have written up a large post about this potential functionality in https://github.com/kube-rs/kube/issues/1148 . Comments would be appreciated.

Reply:
thinking about this, it is probably going to be possible to extend this PR to handle subset streams (if we want to):

- the reflector map here is private and our keying system is private (and can thus be changed easily)
- the keying system could rely on checking subset relations on `ListParams` before calling `setup_reflector`
- the subscribe system could call a potential filter method if `ListParams::is_subset_of`

so don't worry too much about that __mostly disjoint__ problem.